Fog of War Blog Post (#281)
remix-run/remix-website: data/posts/fog-of-war.md

---
title: Fog of War
summary: The Fog of War concept in Remix and React Router provides infinite scalability for your applications
featured: false
draft: true
date: 2024-07-10
image: /blog-images/headers/fog-of-war.png
imageAlt: Fog of War
imageDisableOverlay: true
authors:
  - Matt Brophy
---

Remix is built to ensure your applications are performant by default. With [Fog of War][remix-fog-of-war][^1], we're ensuring this is the case no matter the size of your application.

[^1]: The Fog of War feature was released behind an unstable flag in Remix [v2.10][remix-2-10] for early beta testing - we hope to stabilize it in an upcoming release.

> ## Background

Author: `How Remix makes fetch(es) happen`

> ## Background

Author: On board with "How Remix Fetches Resources" if thats too cheeky though haha

> ## Background

Reviewer: ![image](https://github.com/remix-run/remix-website/assets/12396812/210ac22f-c8f7-45e4-b9a3-04cbfc3ca486)

> ## Background

Reviewer: That's great though, I love it!

Just need to capitalize the words for consistency

Remix has always been mostly a compiler and server-runtime on top of React Router aimed at giving you the idiomatic and performant way we'd have written a React Router SSR app. Could you build your own React Router SSR application without using Remix? Absolutely! However, to get the same kind of performance optimizations, you'd very likely end up writing your own compiler and server-runtime mimicking a lot of the optimizations Remix has built-in.

Most of these Remix optimizations share a common goal of eliminating network waterfalls.

In order to avoid "render then fetch" waterfalls, Remix [decouples rendering from fetching][when-to-fetch] (see also: [Remixing React Router][remixing-rr]). In order to do this, Remix needs to know your route tree up front so that when a link is clicked - it can kick off the data fetches and download the route modules in parallel. This results in an inverted and more performant approach of "fetch then render" (or "fetch while render" if you're [streaming data][streaming]).

> In a "render then fetch" world, your application needs to download the route implementation, then render the component, which would kick off the data fetches - causing a waterfall:

Author: I think it's fair to say that fetching during render (i.e., from a `useEffect`) causes a waterfall?

> In a "render then fetch" world, your application needs to download the route implementation, then render the component, which would kick off the data fetches - causing a waterfall:

Reviewer: Ah yeah, I think that's more correct. I was thinking based on the sentence this was the waterfall:

```
--download route
-----------------render
-----------------------fetch data
```

But I think you're right that it's once the data starts fetching that the waterfall has been created. Either way, I'm probably being too pendatic

Also, need an em dash here 😉 

```suggestion
In a "render then fetch" world, your application needs to download the route implementation, then render the component, which would kick off the data fetches—causing a waterfall:
```

> In a "render then fetch" world, your application needs to download the route implementation, then render the component, which would kick off the data fetches - causing a waterfall:

Author: I reworded a bit to try to clarify as well 👍 

```
In a "render then fetch" world, your application downloads the route implementation, then kicks off the data fetches while rendering the component—causing a waterfall:
```

<img alt="Render then Fetch network diagram" src="/blog-images/posts/fog-of-war/render-then-fetch.png" class="m-auto w-4/5 border rounded-md shadow" />

With "fetch then render", the module fetch and data fetch can be parallelized:

<img alt="Fetch then Render network diagram" src="/blog-images/posts/fog-of-war/fetch-then-render.png" class="m-auto w-4/5 border rounded-md shadow" />

You can take this one step further in Remix via [`<Link prefetch>`][link-prefetch], which allows you to prefetch the route data and components before a user even clicks a link. That way, when the link is clicked the navigation can be instantaneous:

<img alt="Prefetching network diagram" src="/blog-images/posts/fog-of-war/prefetch.png" class="m-auto w-4/5 border rounded-md shadow" />

## The Route Manifest

Let's define a few important terms for clarity:

- **Route tree**: A tree of routes which define the URLs your app can match via parent/child relationships

> - **Route definition**: Aspects of the route used to match a URL - `path`, `index`, `children`

Author: Parts?

> - **Route definition**: Aspects of the route used to match a URL - `path`, `index`, `children`

Reviewer: Yeah I that works. Slight additional tweak to make these two items more consistent

```
- Route definition: Parts of the route used to match a URL (path, index, children)
- Route implementation: Parts of the route needed to load data and render the UI (loader, Component, ErrorBoundary, etc.)
```

And, if we combine this eager discovery with the `<Link prefetch>` optimization above - we can still achieve instantaneous navigation!

<img alt="Fog of War network diagram with eager discovery" src="/blog-images/posts/fog-of-war/fow-eager-prefetch.png" class="m-auto w-4/5 border rounded-md shadow" />

It's also worth noting that because this is all just an optimization, the application works fine without it - just a bit slower because of the network waterfall. So if a user _does_ click that link within the short amount of time it takes to patch it into the manifest, then that link navigation will encounter the waterfall. This is like like `<Link prefetch>` where if the prefetch doesn't complete in time, the fetch happens on click and the user sees a spinner during the navigation. It's also worth noting that a route only has to be discovered once per session. subsequent navigations to the same route won't require a discovery step.

## Visual Explanation

> Lets take a step back and see how this looks from a more visual "route tree" standpoint. Let's look at the current state" of Remix today, where the full manifest is shipped on initial load. In the below route tree, the red dots are the actively rendered route, and the white area conveys that routes the manifest knows about, which is all of them in the current state:

Reviewer: This might not be how you're thinking about it exactly, but it sounded like you were saying the white _is_ the routes, whereas I take the black dots to be the routes

To do this, Remix ships a **Route Manifest** to the client that contains all of your route definitions plus a small amount of metadata. This manifest allows Remix to build a client-side route tree without including the underlying route implementations. Those implementations are loaded via [route.lazy][route-lazy] when a route is navigated to.

Over time, as the user navigates around - more and more route implementations are downloaded via `route.lazy` and the route tree becomes more and more complete.

This approach works for most applications - since the manifest is pretty lightweight and compresses well since it's a repetitive key/value JSON structure. For example, the manifest for https://remix.run/ contains 50 routes, weighs 19.6Kb, but only sends 2.6Kb over the wire after compression.

However, Remix doesn't only want to to provide good performance for small-to-medium sized applications - we want _large_ and _extremely large_ applications to be fast by default too!

> Thankfully, we love dog-fooding React Router and Remix on the myriad of applications at Shopify, both internal and public-facing. As we began rolling out Remix to https://www.shopify.com we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more) - the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module - and thus duplicating the module information (it's path, it's other module `imports`, etc.). This resulted in a manifest that was over 10Mb uncompressed, and ~85Kb compressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.

Reviewer: Take it or leave it, but "thankfully" might sound a bit too pat-ourselves-on-the-back. I was trying to consider a more take-responsibility approach, but happy for you to go whatever direction you like.

```suggestion
Because we love dog-fooding React Router and Remix on the myriad of applications at Shopify, both internal and public-facing, we knew our current strategy was not cutting it for those larger applications. As we began rolling out Remix to https://www.shopify.com we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more) - the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module - and thus duplicating the module information (it's path, it's other module `imports`, etc.). This resulted in a manifest that was over 10Mb uncompressed, and ~85Kb compressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.
```

## Visual Explanation

Lets take a step back and see how this looks from a more visual "route tree" standpoint. Let's look at the current state" of Remix today, where the full manifest is shipped on initial load.

In the below route tree, the red dots are the actively rendered route, and the white area conveys that routes the manifest knows about (all of them in the current state):

<img alt="Route tree showing the entire manifest without Fog of War enabled" src="/blog-images/posts/fog-of-war/tree-1.png" class="m-auto w-4/5 border rounded-md shadow" />

Now, if we enable Fog of War, we'll only ship the active routes in the manifest on initial load:

<img alt="Route tree showing the initial manifest with Fog of War enabled" src="/blog-images/posts/fog-of-war/tree-2.png" class="m-auto w-4/5 border rounded-md shadow" />

> As we hydrate (render) the UI client side, we'll encounter a handful of outgoing link to other routes that are currently unknown to the manifest:

Reviewer: ```suggestion
As we hydrate (render) the UI client side, we'll encounter a handful of links to other routes that are currently unknown to the manifest:
```

<img alt="Route tree showing destination links rendered on the current page" src="/blog-images/posts/fog-of-war/tree-3.png" class="m-auto w-4/5 border rounded-md shadow" />

Remix will discover those routes via a `fetch` call to the Remix server and patch them into them manifest:

<img alt="Route tree with expanded manifest including destination links" src="/blog-images/posts/fog-of-war/tree-4.png" class="m-auto w-4/5 border rounded-md shadow" />

> As you can see - this type of "discovery" allows for the route manifest to start small and grow with the user's path through the app, thus allowing your app to scale to any number of routes without incurring a performance hit on app load.

Reviewer: ```suggestion
As you can see - this type of "discovery" allows for the route manifest to start small and grow with the user's path through the app, thus allowing your app to scale to any number of routes without incurring a performance hit on the app's initial load.
```

> As mentioned earlier, we've been dog-fooding this on https://shopify.com and we're loving the results. Prior to Fog of War, their route manifest contained **1300 routes** and weighted over **10MB uncompressed**. Once Fog of War was enabled, their initial homepage manifest dropped to just **3 routes** and **1.9Kb uncompressed**.

Reviewer: Worth mentioning that it's also on Remix.run too?

## React Router Implementation

> As with most routing features in Remix, it's all React Router under the hood. Fog of War is made possible by a new [`unstable_patchRoutesOnMiss`][rr-patch-routes-on-miss] API. This API allows you to provide an implementation to add new routes to the route tree anytime React Router is unable to match a path in the current route tree. You can implement any async logic you need in this method to discover the appropriate routes and patch them into the current tree any any location you need.

Reviewer: ```suggestion
As with most routing features in Remix, it's all React Router under the hood. Fog of War is made possible by a new [`unstable_patchRoutesOnMiss`][rr-patch-routes-on-miss] API. This API allows you to provide an implementation to add new routes to the route tree anytime React Router is unable to match a path in the current route tree. You can implement any async logic you want in this method to discover the appropriate routes and patch them into the current tree at any location you need.
```

      if (path === "/a") {
        let route = await getARoute(); // { path: 'a', Component: A }
        // Patch the `a` route into the tree as a child of the `root` route
        patch("root", [route]);
      }
    },
  },
);
```

You can expand on this async logic and move towards a manifest-like approach, not too different from what Remix uses, but without the server-aspect:

```js
// Manifest mapping route prefixes to sub-app implementations
let manifest = {
  account: () => await import("./account"),
  dashboard: () => await import("./dashboard"),
};

let router = createBrowserRouter(
  [
    {
      path: "/",
      Component: Home,
    },
  ],
  {
    async unstable_patchRoutesOnMiss({ path, patch }) {
      let prefix = Object.keys(manifest).find((prefix) => path.startsWith(`/${prefix}`));
      if (prefix) {
        let children = await manifest[prefix]();
        patch(null, children);
      }
    },
  }
);
```

This ability to implement async logic also lends itself very well to Micro Frontend and Module Federation architectures in React Router, since you now have an async insertion point to load sub-portions of your application.

> We'd like to give a huge shout out to [Shane Walker][twitter-swalker326] for working with us during the initial release to put together a [great example][rr-mf-example] of using this new API in a federated `rsbuild` React Router application. Make sure to give it a look if you're interested in using Module Federation in your React Router app!

Reviewer: ```suggestion
We'd like to give a huge shout-out to [Shane Walker][twitter-swalker326] for working with us during the initial release to put together a [great example][rr-mf-example] of using this new API in a federated `rsbuild` React Router application. Make sure to give it a look if you're interested in using Module Federation in your React Router app!
```

---
title: Fog of War
summary: Introducing Fog of War - infinitely scalable Remix and React Router applications
featured: false
date: 2024-07-10
image: /blog-images/headers/fog-of-war.png
imageAlt: Fog of War
imageDisableOverlay: true
authors:
  - Matt Brophy
---

Remix is designed to make your application performant by default. Our latest feature, [Fog of War][remix-fog-of-war][^1], ensures your application stays performant no matter how much it grows.

[^1]: The Fog of War feature was released behind an `unstable` flag in Remix [v2.10][remix-2-10] for early beta testing—we hope to stabilize it in an upcoming release

## How Remix makes fetch(es) happen

Remix has primarily been a compiler and server-runtime on top of React Router aimed at giving you the idiomatic and performant way we would have written a React Router SSR app. Could you build your own React Router SSR application without using Remix? Absolutely! However, to get the same kind of performance optimizations, you'd very likely end up writing your own compiler and server-runtime, mimicking a lot of the optimizations Remix has built-in.

> Most of these Remix optimizations share a common goal of eliminating network waterfalls.

Reviewer 2: ```suggestion
Most of Remix's optimizations share a common goal of eliminating network waterfalls.
```

In order to avoid "render then fetch" waterfalls, Remix [decouples rendering from fetching][when-to-fetch] (see also: [Remixing React Router][remixing-rr]). In order to do this, Remix needs to know your route tree up front so it can kick off the data fetches and download the route modules in parallel whenever a link is clicked. This results in the inverted and more performant approach of "fetch then render" (or "fetch while render" if you're [streaming data][streaming]).

> In a "render then fetch" world, your application needs to download the route implementation, then render the component, which would kick off the data fetches - causing a waterfall:

Reviewer 2: ```suggestion
In a "render then fetch" world your application needs to download the route implementation, then render the component, which would kick off data fetching and cause a waterfall.
```

Author: I had reworded this slightly above to clarify based on Brook's feedback - does this new wording seem clear to you?

```
In a "render then fetch" world, your application downloads the route implementation, then kicks off the data fetches while rendering the component—causing a waterfall:
```

Reviewer: Sounds good to me! Thanks Broph Dawg

<img alt="Render then Fetch network diagram" src="/blog-images/posts/fog-of-war/render-then-fetch.png" class="m-auto w-4/5 border rounded-md shadow" />

With "fetch then render", the module fetch and data fetch can be parallelized:

<img alt="Fetch then Render network diagram" src="/blog-images/posts/fog-of-war/fetch-then-render.png" class="m-auto w-4/5 border rounded-md shadow" />

You can take this one step further in Remix via [`<Link prefetch>`][link-prefetch], which allows you to prefetch the route data and components before a user even clicks a link. That way, when the link is clicked the navigation can be instantaneous:

<img alt="Prefetching network diagram" src="/blog-images/posts/fog-of-war/prefetch.png" class="m-auto w-4/5 border rounded-md shadow" />

## The Route Manifest

> Let's define a few important terms for clarity:

Reviewer 2: ```suggestion
First of all, let's define a few important terms for clarity:
```

  "hasClientAction": false,
  "hasClientLoader": false,
  "hasErrorBoundary": true,
  "module": "/assets/root-x1zXK6d6.js",
  "imports": [
    "/assets/entry.client-uo5Ucqv5.js",
    "/assets/utils-c6MmN8mv.js",
    "/assets/color-scheme-y8FjcTs4.js",
    "/assets/icons-6-7TeyZS.js"
  ],
  "css": [
    "/assets/root-Wq_jPy7B.css"
  ]
}
```

Over time, as the user navigates around—more and more route implementations are downloaded via `route.lazy` and the route tree becomes more and more complete.

This approach works for most applications - since the manifest is pretty lightweight and compresses well since it's a repetitive key/value JSON structure. For example, the manifest for https://remix.run/ contains 50 routes, weighs 19.6Kb uncompressed, but only sends 2.6Kb over the wire after compression.

However, Remix doesn't only want to to provide good performance for small-to-medium sized applications—we want _large_ and _extremely large_ applications to be fast by default too! Because we love dog-fooding React Router and Remix on the myriad of applications at Shopify, both internal and public-facing, we knew our current strategy was not cutting it for those larger applications.

> As we began rolling out Remix to https://www.shopify.com, we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more)–the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module, and thus duplicating the module information (it's path, it's other module `imports`, etc.). This resulted in a manifest that was ~85Kb compressed, and ~10Mb uncompressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.

Reviewer 3: ```suggestion
As we began rolling out Remix to https://www.shopify.com, we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more)–the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module, and thus duplicating the module information (its path, its other module `imports`, etc.). This resulted in a manifest that was ~85Kb compressed, and ~10Mb uncompressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.
```

## Fog of War

> At Remix, we're big fans of the Retro vibes of our younger years: from old school web development using HTML `<form>` elements and HTTP `POST` requests, to 90's music, to retro video games with [ever expanding maps][wikipedia-fog-of-war]. These expanding game maps provided the inspiration for (at the very least) the name of our solution to this problem of ever-growing route manifests.

Reviewer 3: ```suggestion
At Remix, we're big fans of the Retro vibes of our younger years: from old school web development using HTML `<form>` elements and HTTP `POST` requests, to '90s music, to retro video games with [ever expanding maps][wikipedia-fog-of-war]. These expanding game maps provided the inspiration for (at the very least) the name of our solution to this problem of ever-growing route manifests.
```
